File: vehicle-tracker-backend/vehicle-service/src/main/java/com/github/diogocerqueiralima/domain/model/Device.java
package com.github.diogocerqueiralima.domain.model;

import java.time.Instant;
import java.util.Objects;
import java.util.UUID;

public class Device {

    private final UUID id;
    private final String imei;
    private final String serialNumber;
    private final String manufacturer;
    private final Vehicle vehicle;

    public Device(UUID id, String imei, String serialNumber, String manufacturer, Vehicle vehicle) {
        this.id = id;
        this.imei = imei;
        this.serialNumber = serialNumber;
        this.manufacturer = manufacturer;
        this.vehicle = vehicle;
    }

    public Device(String imei, String serialNumber, String manufacturer, Vehicle vehicle) {
        this(null, imei, serialNumber, manufacturer, vehicle);
    }

    public UUID getId() {
        return id;
    }

    public String getImei() {
        return imei;
    }

    public String getSerialNumber() {
        return serialNumber;
    }

    public String getManufacturer() {
        return manufacturer;
    }

    public Vehicle getVehicle() {
        return vehicle;
    }

    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) return false;
        Device device = (Device) o;
        return Objects.equals(imei, device.imei) && Objects.equals(serialNumber, device.serialNumber);
    }

    @Override
    public int hashCode() {
        return Objects.hash(imei, serialNumber);
    }

}

File: vehicle-tracker-mobile/app/src/main/java/com/github/diogocerqueiralima/domain/model/Device.kt
package com.github.diogocerqueiralima.domain.model

import java.util.UUID
import kotlin.time.Instant

/**
 * Represents a tracking device asset with its identifying and manufacturing attributes.
 */
data class Device(
    val id: UUID,
    val createdAt: Instant,
    val updatedAt: Instant,
    val serialNumber: String,
    val model: String,
    val manufacturer: String,
    val imei: String
) {

    init {

        // 1. Serial number, model and manufacturer are required identifying attributes.
        require(serialNumber.isNotBlank()) { "serialNumber must not be blank" }
        require(model.isNotBlank()) { "model must not be blank" }
        require(manufacturer.isNotBlank()) { "manufacturer must not be blank" }

        // 2. IMEI must follow the standard 15-digit format.
        require(imei.isNotBlank()) { "imei must not be blank" }
        require(imei.length == 15) { "imei must be 15 characters long" }

    }

    /**
     * Human-readable name composed of the manufacturer and model.
     */
    val displayName = "$manufacturer $model"

}
